tensorflow/compiler/mlir/xla/transforms/lhlo_legalize_to_affine.cc

```diff
@@ -17,20 +17,74 @@ limitations under the License.
 
 #include "absl/memory/memory.h"
 #include "mlir/Dialect/Affine/IR/AffineOps.h"  // from @llvm-project
-#include "mlir/Dialect/StandardOps/IR/Ops.h"  // from @llvm-project
-#include "mlir/IR/Attributes.h"  // from @llvm-project
-#include "mlir/IR/Location.h"  // from @llvm-project
-#include "mlir/IR/MLIRContext.h"  // from @llvm-project
-#include "mlir/IR/PatternMatch.h"  // from @llvm-project
-#include "mlir/IR/StandardTypes.h"  // from @llvm-project
-#include "mlir/Pass/Pass.h"  // from @llvm-project
+#include "mlir/Dialect/StandardOps/IR/Ops.h"   // from @llvm-project
+#include "mlir/IR/Attributes.h"                // from @llvm-project
+#include "mlir/IR/Location.h"                  // from @llvm-project
+#include "mlir/IR/MLIRContext.h"               // from @llvm-project
+#include "mlir/IR/PatternMatch.h"              // from @llvm-project
+#include "mlir/IR/StandardTypes.h"             // from @llvm-project
+#include "mlir/Pass/Pass.h"                    // from @llvm-project
 #include "tensorflow/compiler/mlir/xla/ir/lhlo_ops.h"
 #include "tensorflow/compiler/mlir/xla/transforms/map_xla_to_scalar_op.h"
 
 namespace mlir {
 namespace xla_lhlo {
 namespace {
 
+struct DotOpConverter : public OpRewritePattern<DotOp> {
+  using OpRewritePattern<DotOp>::OpRewritePattern;
+  // Supports only rank-2 tensors.
+  LogicalResult matchAndRewrite(DotOp op,
+                                PatternRewriter& rewriter) const override {
+    const auto& lhs = op.lhs();
+    const auto& rhs = op.rhs();
+    const auto& lhs_type = lhs.getType().template cast<MemRefType>();
+    const auto& rhs_type = rhs.getType().template cast<MemRefType>();
+    const auto& element_type = lhs_type.getElementType();
+    const auto& shape_lhs = lhs_type.getShape();
+    const auto& shape_rhs = rhs_type.getShape();
+
+    if ((lhs_type.getRank() != 2) || (rhs_type.getRank() != 2) ||
+        (shape_lhs.back() != shape_rhs.front())) {
+      return failure();
+    }
+    SmallVector<Value, 4> lhs_indices;
+    SmallVector<Value, 4> rhs_indices;
+    SmallVector<Value, 4> result_indices;
+    const auto& loc = op.getLoc();
+
+    auto forOp = rewriter.create<AffineForOp>(loc, 0, shape_lhs[0]);
+    lhs_indices.push_back(forOp.getInductionVar());
+    result_indices.push_back(forOp.getInductionVar());
+    rewriter.setInsertionPointToStart(forOp.getBody());
+
+    for (int i = shape_rhs.size() - 1; i >= 0; --i) {
```

Comment:
It's actually simpler and clearer to write this as:
```
   rhs_indices.resize(2);
   auto forOp = rewriter.create<AffineForOp>(loc, 0, shape_rhs.back());
   result_indices[1] = forOp.getInductionVar());
   rhs_indices.push_back(forOp.getInductionVar());
   rewriter.setInsertionPointToStart(forOp.getBody());
   forOp = rewriter.create<AffineForOp>(loc, 0, shape_rhs[i]);
   lhs_indices.push_back(forOp.getInductionVar());
   rhs_indices[0] = forOp.getInductionVar());
```
No need of reverse on rhs_indices now.